File: lib/features/diet/domain/nutrition/planned_meal_log.dart
import '../meal.dart';
import 'food_log_entry.dart';
import 'food_reference.dart';

/// Turns a planned meal into food-log entries.
///
/// This is what makes ticking a meal honest. The interaction stays a single
/// tap, but instead of writing "meal X was eaten" it writes what the plan says
/// that meal WAS — one entry per item, each carrying the plan's own figures,
/// its `estimated` provenance, and a link back to the meal so un-ticking can
/// remove exactly these and nothing else.
///
/// The figures come from the plan document, not the nutrition catalog, and are
/// labelled [NutritionSource.dietPlan] to say so. For an imported plan they
/// were AI-estimated at import time; that flag rides along rather than being
/// laundered into a measurement.
///
/// Pure: [now] and [idPrefix] are injected so the result is deterministic.
List<FoodLogEntry> entriesForPlannedMeal({
  required Meal meal,
  required DateTime day,
  required DateTime now,
  required String idPrefix,
}) {
  final entries = <FoodLogEntry>[];
  for (var i = 0; i < meal.items.length; i++) {
    final item = meal.items[i];
    // An item with no calorie figure contributes nothing countable. It is
    // still logged, at zero, so the day's list shows everything the user
    // ticked rather than quietly dropping the parts the plan never costed.
    entries.add(
      FoodLogEntry(
        id: '$idPrefix-$i',
        day: DateTime(day.year, day.month, day.day),
        loggedAt: now,
        foodId: 'plan:${meal.id}#$i',
        foodName: item.name,
        quantity: item.quantity,
        unit: item.unit,
        // The plan states a quantity in its own unit; it does not state a
        // mass, and inferring one would be a guess. Zero means "not weighed".
        grams: item.unit.toLowerCase() == 'g' ? item.quantity : 0,
        kcal: item.calories ?? 0,
        proteinG: item.proteinG ?? 0,
        carbsG: item.carbsG ?? 0,
        fatG: item.fatG ?? 0,
        source: NutritionSource.dietPlan,
        sourceRef: '${meal.id}#$i',
        origin: FoodLogOrigin.plannedMeal,
        estimated: item.estimated,
        mealId: meal.id,
      ),
    );
  }
  return entries;
}

File: functions/diet/planned_meal_log.js
/**
 * The server's mirror of `lib/features/diet/domain/nutrition/
 * planned_meal_log.dart` (`entriesForPlannedMeal`): ticking a planned meal
 * writes what the plan says that meal WAS — one food-log entry per item, with
 * the plan's own figures, its `estimated` provenance, and a link back to the
 * meal so un-ticking removes exactly these and nothing else.
 *
 * Until this existed a meal ticked from Ask wrote only its `dietEntries` tick,
 * so on any day that also had a logged food the meal silently dropped out of
 * "what was eaten" (consumption reads the log whenever it has anything). The
 * ids and fields match the app's byte for byte, so a meal ticked on either
 * side and un-ticked on the other still removes cleanly.
 *
 * Pure: the caller writes the rows.
 */

/**
 * @param {!Object} meal A plan Meal (`{id, items}`).
 * @param {string} dayKey 'yyyy-MM-dd'
 * @return {!Array<!Object>} Rows in `store.writeFoodLog`'s shape.
 */
function entriesForPlannedMeal(meal, dayKey) {
  return (meal.items || []).map((item, i) => ({
    id: `${dayKey}__${meal.id}-${i}`,
    dayKey,
    foodId: `plan:${meal.id}#${i}`,
    foodName: item.name,
    quantity: item.quantity,
    unit: item.unit,
    // The plan states a quantity in its own unit, not a mass; zero means
    // "not weighed" rather than a guessed conversion.
    grams: String(item.unit || "").toLowerCase() === "g" ? item.quantity : 0,
    kcal: typeof item.calories === "number" ? item.calories : 0,
    proteinG: typeof item.proteinG === "number" ? item.proteinG : 0,
    carbsG: typeof item.carbsG === "number" ? item.carbsG : 0,
    fatG: typeof item.fatG === "number" ? item.fatG : 0,
    source: "dietPlan",
    sourceRef: `${meal.id}#${i}`,
    origin: "plannedMeal",
    estimated: item.estimated === true,
    mealId: meal.id,
  }));
}

module.exports = {entriesForPlannedMeal};
